src/main/java/hudson/tasks/test/AbstractTestResultAction.java

```diff
@@ -355,19 +668,573 @@ private CategoryDataset buildDataSet(StaplerRequest req) {
+                Mark that the testcase passed after failing.
+                 */
+                Integer index = testCaseIndex.get(caseName);
+                if (index != null) {
+                    List<Integer> infoList = testInfo.get(index);
+                    infoList.set(2, 1);
+                }
+            }
+            /*
+            Recording the build number examined in a separate data series.
+             */
+            xySeries.add(testResultAction.run.number, null);
+            /*
+            Queuing the build number along with all the testcases that failed in that build.
+             */
+            buildHistory.add(new Pair<>(testResultAction, buildSet));
+        }
+        /*
+        Calling utility method to calculate the number of test flappers in a build.
+         */
+        while (!buildHistory.isEmpty()) {
+            shiftBuildHistoryUtil(buildHistory, testsHistory);
+        }
+        /*
+        List containing all the testcases along with number of times they flapped or number of time they failed. Sorting
+        that list and selecting top testcases for display on the chart.
+         */
+        List<Pair<Integer, String>> testFailCount = new ArrayList<>();
+        for (String caseName : testCaseIndex.keySet()) {
+            int index = testCaseIndex.get(caseName);
+            if (orderBy.equals(FLAPMETRIC)) {
+                testFailCount.add(new Pair<Integer, String>(testInfo.get(index).get(1), caseName));
+            }
+            else {
+                testFailCount.add(new Pair<Integer, String>(testInfo.get(index).get(0), caseName));
+            }
+        }
+        testFailCount.sort(new PairComparator<Integer, String>());
+        testsToDisplay = Math.min(testsToDisplay, testFailCount.size());
+        /*
+        A map storing the testcase along with number of times it flapped and number of times it failed for displaying as
+        tool tip.
+         */
+        flapperInfo = new HashMap<>();
+        /*
+        Data structure for indexing all the testcases which will be displayed on the chart and this data structure will
+        also be used for generating tool tips.
+         */
+        mostFailedTestCases = new ArrayList<>(testsToDisplay);
+        xySeries.add(this.run.number + 0.5, testsToDisplay + 0.5);
+        failSeries.add(xySeries);
+        for (int i = 1; i <= testsToDisplay; i++) {
+            xySeries = new XYSeries(i);
+            mostFailedTestCases.add("");
+            failSeries.add(xySeries);
+        }
+        /*
+        Populating data structures used by JFreeChart API for displaying trends and generating tool tips.
+         */
+        for (int testIndex = 1; testIndex <= testsToDisplay; testIndex++) {
+            Pair<Integer, String> testCase = testFailCount.get(testIndex - 1);
+            String caseName = testCase.second;
+            int y = testsToDisplay - testIndex + 1;
+            int index = testCaseIndex.get(caseName);
```

Comment:
Try to minimize the var declaration if its not helpful at all places.

Reply:
Done.